packages/react-components/react-motion-preview/src/hooks/useMotion.ts

```diff
@@ -0,0 +1,484 @@
+  readonly unit: string;
+}
+
+/**
+ * Style property map read only.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/StylePropertyMapReadOnly
+ */
+interface StylePropertyMapReadOnly {
+  [Symbol.iterator](): IterableIterator<[string, CSSUnitValue[]]>;
+
+  get(property: string): CSSUnitValue | undefined;
+  getAll(property: string): CSSUnitValue[];
+  has(property: string): boolean;
+  readonly size: number;
+}
+
+/**
+ * HTMLElement with styled map.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Element/computedStyleMap
+ */
+type HTMLElementWithStyledMap<T extends HTMLElement = HTMLElement> = T & {
+  computedStyleMap(): StylePropertyMapReadOnly;
+};
+
+/**
+ * CSS with number parsing.
+ * @see https://drafts.css-houdini.org/css-typed-om-1/#css
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/CSS/number
+ */
+type CSSWithNumber = typeof CSS & {
+  number(value: number): {
+    value: number;
+    readonly unit: string;
+  };
+};
+
+/**
+ * @internal
+ *
+ * Motion props.
+ */
+export type MotionProps<T = HTMLElement> = {
+  /**
+   * Whether the element should be present in the DOM.
+   *
+   * @default false
+   */
+  presence: boolean;
+
+  /**
+   * Ref to the element.
+   *
+   * @example
+   * const motion = useMotion<HTMLDivElement>({ presence: isOpen });
+   *
+   * <div ref={motion.ref} />
+   */
+  ref?: RefObjectFunction<T> | React.RefCallback<T> | React.Ref<T>;
+
+  /**
+   * Whether the element is currently active in the DOM.
+   * Useful to apply CSS transitions only when the element is active.
+   *
+   * @example
+   * const motion = useMotion<HTMLDivElement>({ presence: isOpen });
+   *
+   * <div ref={motion.ref} className={motion.active ? 'element element-active' : 'element'} />
+   */
+  active?: boolean;
+
+  /**
+   * Current state of the element.
+   *
+   * - `unmounted` - The element is not yet rendered or can be safely removed from the DOM.
+   * - `entering` - The element is performing enter animation.
+   * - `entered` - The element has finished enter animation.
+   * - `idle` - The element is currently not animating, but rendered on screen.
+   * - `exiting` - The element is performing exit animation.
+   * - `exited` - The element has finished exit animation.
+   *
+   * @example
+   * const motion = useMotion<HTMLDivElement>({ presence: isOpen });
+   *
+   * <div ref={motion.ref} className={`element element-${motion.state}`} />
+   */
+  state?: 'unmounted' | 'entering' | 'entered' | 'idle' | 'exiting' | 'exited';
+};
+
+export type MotionOptions = {
+  /**
+   * Whether to animate the element on first mount.
+   *
+   * @default false
+   */
+  animateOnFirstMount?: boolean;
+};
+
+/**
+ * @internal
+ *
+ * Returns CSS styles of the given node.
+ * @param node - DOM node.
+ * @returns - CSS styles.
+ */
+const getElementComputedStyle = (node: HTMLElement): CSSStyleDeclaration => {
+  const win = node.ownerDocument?.defaultView ? node.ownerDocument.defaultView : window;
```

Comment:
Can you add a Github issue to update this to remove direct access to `window` in the future?

We'll be disallowing access to `window` via eslint: https://github.com/microsoft/fluentui/pull/28887

I would say change it now, but it's still an open question as to what we'll do instead.

Reply:
Can be an exception? We are checking if we are in DOM env using `canUseDOM` and we necessarily need to use window methods in case that is true.

Reply:
It's not that you cannot use `window` APIs, the issue is using the right `window`. In the case of `canUseDOM` I don't think it matters which `window` object we reference as I think we'd expect `window` to either exist in the case of client side execution or not in the case of SSR and we're not using any of its APIs to do anything, just checking their existence.

The latter case where we use `window` matters because if you access `window` APIs directly you may be referencing the wrong window. For example, if you pop a window out in some apps this can happen and it leads to issues with styling, etc. This has happened in multiple apps where something like `document.createElement('div')` is called from a popout window (e.g., a window made with `window.open()`) but the `document` (and therefore `window`) element is actually the parent window (e.g., the window that called `window.open()`).

---

If you think changing this will be a breaking change, let's hold off on merging this until we settle on a way forward. At present there are two things we can do:

1. Use `useWindow/useDocument` from [`react-window-provider`](https://github.com/microsoft/fluentui/tree/master/packages/react-window-provider). This is a FUIv8-era package though so I'm not sure we want to use it for v9. Note: it doesn't bring in any v8 deps, it's just an older package.
2. Use `useFluent()` from [`react-shared-contexts`](https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-shared-contexts/src/ProviderContext/ProviderContext.ts#L28) to get window via `document.defaultView`.

We could also consider adding an new API to `FluentProvider` to explicitly set `window` like we do for `targetDocument`.

Regardless of what we settle on rather than directly access `window` we'll instead do something like:

```
// ❌ Not this
window.someAPI();
// ✅ This via react-window-provider
const win = useWindow();
win.someAPI();
// ✅ Or this via useFluent()
const win = useFluent().targetDocument.defaultView;
win.someAPI();

```

and then use `win` just like we'd use `window`.

Reply:
I don't think it's a breaking change and we `can refactor this. The issue is that the function using window APIs is being used outside of React components, so a hook-based solution through context won't work here. If we have a `getWindow` function available somewhere, I can make the change. 

The current code only calls the "default" window object in case we can use DOM and only if we cannot find it from the current referenced element (by using `node.ownerDocument.defaultView`), which is the way that other parts of V9 codebase are using, including the `useFluent` solution.